### PR TITLE
Gallery Block: Add random order setting

### LINKED ISSUE #9701: Gallery Block: Missing 'Random Order' setting
In the classic editor, when I go to set up a WordPress image gallery, there are "Gallery Settings" for "Link To", number of "Columns", "Random Order", "Size", and "Type" ("Thumbnail Grid", "Slideshow"). Those settings are missing in Gutenberg Version 3.7.0 when using the "Gallery" block. 

Related: The captions on the Gutenberg "Gallery" block are styled so that the caption hangs oddly over the image. The images also can't be clicked on (possibly because there is no "Link To" option, as mentioned). 

**To Reproduce**
1. Use the 'Gallery' block 
2. Click on 'Media Library'
3. Select Images
4. Click on 'Create a new gallery'
5. Now you are on the final gallery set-up page. Above the 'Insert Gallery' button is a blank area—in that blank area is where the settings I mentioned would normally appear in the classic editor. 

**Expected behavior**
I expected the Gutenberg "Gallery" block to have "Gallery Settings" so that I'd have control over the following options:  "Link To", number of "Columns", "Random Order", "Size", and "Type" ("Thumbnail Grid", "Slideshow").  

**Screenshots**
screen shot 2018-09-07 at 1 15 37 pm
screen shot 2018-09-07 at 1 16 27 pm
screen shot 2018-09-07 at 1 17 34 pm

**Desktop (please complete the following information):**
 - OS: iOS
 - Browser: Firefox
 - Version: 61.0.2

**Additional Context**
 - I'm on the Genesis framework

### PR DESCRIPTION
**Pull request by t-hamano**
Fixes #9701

## What?

This PR adds a new "Random order" setting to the Gallery block.

image

## Why?
The gallery shortcode has[ many options](https://codex.wordpress.org/Gallery_Shortcode#Options). The Gallery block substantially covers many of the shortcode options, as we can reorder the images inside. It also automatically inherits some options when pasting a shortcode. However, only random order cannot be achieved unless the additional setting is added.

## How?
In addition to adding settings, I've made it so that the random option is inherited when a shortcode is converted to a block. In the future, we may support other values for `orderby` (`menu_order`, `title`, etc.) and `order` options.

## Testing Instructions
- Paste the shortcode below into the editor. Please change the media ID to one that exists in your environment.
  `[gallery ids="1,2,3,4" orderby="rand"]`
- It should be converted to a gallery block and the Random order setting should be enabled.
- Save the post and reload the browser on the front end.
- The order of the images should change every time you reload the browser.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/54422211/29baa0b9-b028-4df8-b15b-aedf8ffc32ff

### CODE REVIEW

File: packages/block-library/src/gallery/index.php [hunk 1]
```diff
@@ -32,6 +32,24 @@ function block_core_gallery_data_id_backcompatibility( $parsed_block ) {
...
+ *
+ * @param array $parsed_block The block being rendered.
+ * @return array The block object with randomized order of image blocks.
+ */
+function block_core_gallery_random_order( $parsed_block ) {
+	if ( 'core/gallery' === $parsed_block['blockName'] && isset( $parsed_block['attrs']['randomOrder'] ) && $parsed_block['attrs']['randomOrder'] ) {
+		$inner_blocks = $parsed_block['innerBlocks'];
+		shuffle( $inner_blocks );
+		$parsed_block['innerBlocks'] = $inner_blocks;
+	}
```
Comment (aaronrobertshaw) [inline packages/block-library/src/gallery/index.php]:
```suggestion
	if ( 'core/gallery' === $parsed_block['blockName'] && ! empty( $parsed_block['attrs']['randomOrder'] ) ) {
		shuffle( $parsed_block['innerBlocks'] );
	}

	return $parsed_block;
```

We could make this a little less verbose given we're not checking strictly for `true` and are already mutating `$parsed_block`.

What do you think?

**Review by aaronrobertshaw — APPROVED** [review]
Nice work @t-hamano 👍 

This tests really well for me. I left one minor suggestion as an inline comment below but I'll leave that up to you.

✅ Pasted shortcode is successfully transformed to a gallery with random order
✅ Transforming images to galleries and vice versa still works
✅ Manually creating a randomly order gallery works
✅ A randomly ordered gallery does have the order of its images changed on each load
✅ Fixture tests all pass

LGTM 🚢

File: packages/block-library/src/gallery/index.php (hunk 1 above)
Comment (t-hamano) [reply packages/block-library/src/gallery/index.php]: I definitely think it's simpler and easier to understand 👍

**Comment by a4jp-com** [comment]
Thank you. It will be great to have the random setting back again.

**Comment by cagrimmett** [comment]
🙌

**Comment by suzanerostoff** [comment]
Thank you a lot for this feature. Many issues out there request the order by random in one block or another...

Could you please take one second to transport/migrate the solution approved here to solve it's brother issue? -> https://github.com/WordPress/gutenberg/issues/40481

Please please❤️..

**Comment by ghost** [comment]
Thank you a lot for this feature. Many issues out there request the order by random in one block or another...

Could you please take one second to transport/migrate the solution approved here to solve its brother issue? -> https://github.com/WordPress/gutenberg/issues/40481

Please please❤️..

**Pull request merged**